test/test_common/utility.cc

```diff
@@ -222,6 +222,28 @@ AssertionResult TestUtility::waitForGaugeEq(Stats::Store& store, const std::stri
   return AssertionSuccess();
 }
 
+AssertionResult TestUtility::waitUntilHistogramHasSamples(Stats::Store& store,
+                                                          const std::string& name,
+                                                          Event::TestTimeSystem& time_system,
+                                                          std::chrono::milliseconds timeout) {
+  Event::TestTimeSystem::RealTimeBound bound(timeout);
+  while (true) {
+    auto histo = findByName<Stats::ParentHistogramSharedPtr>(store.histograms(), name);
+    if (histo) {
+      if (histo->cumulativeStatistics().sampleCount()) {
```

Comment:
this call is not thread safe and is triggering TSAN errors, please fix ASAP.

https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/78702/logs/189

Reply:
Hmm, how come TSAN didn't complain with the original diff? I'll send a fix in a bit

Reply:
You sure it's me, seeing quic integration failing TSAN too:

https://dev.azure.com/cncf/envoy/_build/results?buildId=78720&view=logs&jobId=d1f76054-8f79-554b-6f4a-11d6a63b8e00&j=d1f76054-8f79-554b-6f4a-11d6a63b8e00&t=266e17e3-d213-54b5-deef-0dcee01da137

Reply:
Well, the integration test I touched did become flaky so maybe it is me:

```
//test/integration:protocol_integration_test                              FLAKY, failed in 1 out of 6 in 227.4s
  Stats over 6 runs: max = 227.4s, min = 214.3s, avg = 219.7s, dev = 5.2s
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-dbg/testlogs/test/integration/protocol_integration_test/shard_5_of_5/test_attempts/attempt_1.log
//test/integration:quic_http_integration_test                            FAILED in 4 out of 8 in 37.9s
  Stats over 8 runs: max = 37.9s, min = 33.8s, avg = 35.5s, dev = 1.3s
```

Looking into it...

Reply:
the quic TSAN failure is here: https://gist.github.com/rgs1/1d352d45a78779b14e68cde4ac39e254

Reply:
Ok now I see mine:

https://gist.github.com/rgs1/bbd811c453484d8887a0448cc8c3d99c

This race is acceptable, so I think we can tell TSAN to ignore it.

Reply:
The TSAN failure involves waitUntilHistogramHasSamples which was introduced in this PR.  That's why I suspect it is related to this.

2021-06-11T18:24:15.8290833Z     #0 Envoy::Stats::HistogramStatisticsImpl::sampleCount() const /proc/self/cwd/./source/common/stats/histogram_impl.h:62:50 (protocol_integration_test+0x7c8b879)
2021-06-11T18:24:15.8292040Z     #1 Envoy::TestUtility::waitUntilHistogramHasSamples(Envoy::Stats::Store&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Event::TestTimeSystem&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >) /proc/self/cwd/test/test_common/utility.cc:233:41 (protocol_integration_test+0x794eef8)

Reply:
Possible fix: https://github.com/envoyproxy/envoy/pull/16946